### DISCUSSION TITLE
Rich-Club Coefficient by Node Rank

### DISCUSSION THREAD

**Discussion by ignRyann** [networkx/networkx#7050] [Algorithms]
### Explanation
Returns the rich-club coefficient of the graph G by node rank.

> Nodes in the network are sorted by decreasing number of links that each node contains. There are instances where groups of nodes contain identical numbers of links. Where this occurs, they are arbitrarily assigned a position within that group. The node rank $r$ denotes the position of a node on this ordered list.
> 
> The rich-club connectivity $\phi (r)$ defined as the ratio of the total number of links to the maximum possible number of links between members of the rich-club. The maximum possible number of links between $n$ nodes is $\frac{n(n-1)}{2}$.
>
> [[The Rich-Club Phenomenon in the Internet Topology by Shi Zhou and Raul J. Mondragon](https://ieeexplore.ieee.org/document/1278314)]

### Parameters
**G: NetworkX graph**
- Undirected graph with neither parallel edges nor self-loops.

### Returns
**rc : dictionary**
- A dictionary, keyed by a node's rank, with rich-club coefficient values.

### Solution
```python
def rich_club_coefficient_by_rank(G):
    def calc_rank_rich_club_coeff(G, ranked_nodes, r):
        # Retrieving nodes in rich-club
        top_r_nodes = ranked_nodes[0 : r]
        
        # Calculating E_r
        node_set = set(top_r_nodes)
        E_r = 0
        for node in node_set:
            neighbors = set(G.neighbors(node))
            edges += len(neighbors.intersection(node_set))

        # Rich-Club Coefficient as a function of Node Rank formula
        return (2 * E_r) / (r * (r - 1))

    # Nodes are sorted by decreasing number of links that each node contains
    ranked_nodes = [node for node, _ in sorted(G.degree(), key=lambda x: x[1], reverse=True)]

    # Calculating for all rank values
    rc = dict([(rank, calc_rich_club_coeff(G, ranked_nodes, rank)) for rank in range(2, G.number_of_nodes() + 1)])

    return rc
```

**Comment by dschult** [comment]
But we already have [`nx.rich_club_coefficient(G, normalized=True, Q=100, seed=None)`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.richclub.rich_club_coefficient.html)
Thanks for thinking of us.

**Reply by ignRyann** [reply]
Doesn't [nx.rich_club_coefficient(G, normalized=True, Q=100, seed=None)](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.richclub.rich_club_coefficient.html) return the rich-club coefficient keyed by degree?

Or maybe, we could an extra argument `mode` to calculate rich-club coefficient as either **(i)** a function of degree, or **(ii)** a function of node rank?

What do you think? Just wanting to hear opinions on the idea or if it's not necessary and why not

**Comment by dschult** [comment]
Ahh... the term `rank` was confusing me.  The rank is the index of the node when sorted by degree.
It's almost like a spearman rank correlation instead of pearson correlation I guess.

I guess my response is: is that really used very much?  If so, then we should probably include a variant (like with a keyword input as you suggest) that returns it.  But I'd want to be convinced that it is used widely and that it gives effectively different results.

**Reply by ignRyann** [reply]
Apologies for the lack of clarification but it is as you've said; we define a node's rank $r$ as its position in a sorted list of decreasing degrees.

I'm currently taking a module that makes use of the NetworkX library and whilst there was a function to aid most of my tasks, I wasn't able to find one for the rich-club coefficient as a function of node rank $r$. Seeing as how it's being taught at top-level institutions, I guess it has some sort of rising relevancy. However, I am no expert in the field but I thought it would be nice to have. The papers that I know that makes use of it are:

- [Rich-Cores in Networks by Athen Ma, Raul J Mondragon](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0119678)
- [The rich-club phenomenon in the Internet topology by Shi Zhou and R. J. Mondragon](http://dx.doi.org/10.1109/LCOMM.2004.823426)

The main difference I guess is that we just focus specifically on the highest-degree nodes rather than looking at any arbitrary degree threshold $k$, which might include a large portion of the nodes. It would also help across networks with different degree distributions as the meaning of a specific degree threshold $k$ might differ. However, the top $r$ nodes always represent the same fraction of highest degree nodes, making the rank-based approach more comparable across diverse networks.

$\phi(r) = \frac{2E(r)}{r(r-1}$

![Rich-Club Coefficient against node rank](https://github.com/networkx/networkx/assets/63001589/8c645f54-22c2-4870-a6c9-2cc376543995)

![Rich-Club Coefficient against node rank (top percent)](https://github.com/networkx/networkx/assets/63001589/cf81ecf7-ac91-4bd4-93b9-f42cf94686c8)

$\phi(k) = \frac{2E_k}{N_k(N_k - 1)}$

![Rich-Club Coefficient against node degree](https://github.com/networkx/networkx/assets/63001589/defc7679-4d64-4f69-8ac4-3e4e24f44092)
